packages/stream-model-instance/src/model-instance-document.ts

```diff
@@ -168,6 +162,31 @@ export class ModelInstanceDocument<T = Record<string, any>> extends Stream {
     )
   }
 
+  /**
+   * Creates a deterministic ModelInstanceDocument with a 'set' accountRelation.
+   * @param ceramic - Instance of CeramicAPI used to communicate with the Ceramic network
+   * @param metadata - Genesis metadata
+   * @param unique - Unique fields values
+   * @param opts - Additional options
+   */
+  static async set<T>(
+    ceramic: CeramicApi,
+    metadata: ModelInstanceDocumentMetadataArgs,
+    unique: Array<string>,
```

Comment:
hmm I'm a bit worried about the arguments being ordering-dependent here.  I'm afraid it will be error-prone to users, and requires them to keep track of what order the fields were specified in the Model definition.  Can we instead make this a dictionary where the user has to specify the field names of the fields that get the uniqueness properties?  I'm fine if it still get encoded in an ordering-dependent way into the commit itself, but for the user-facing APIs I'd like to make them a bit harder to misuse by accident.

Reply:
I don't mind changing it, but fundamentally the logic should be based on the `accountRelation.fields` array of the model anyway, for example as defined currently:

```ts
const unique = model.accountRelation.fields.map(field => doc.content[field] ?? '')
```

Then the handler side just needs to do something like:

```ts
const bytes = fromString(unique.join('|'))
```

Changing it to a dictionary would require the client to still use the `accountRelation.fields` as input anyway, for example:

```ts
const unique = model.accountRelation.fields.reduce((acc, field) => {
  acc[field] = doc.content[field] ?? ''
  return acc
}, {})
```

Then the handler side would need to convert it back to an array based on the model:

```ts
const uniqueParts = model.accountRelation.fields.map(field => {
  const value = unique[field]
  if (value == null) {
    throw new Error(`Missing set field value: ${field}`)
  }
  return value
})
const bytes = fromString(uniqueParts.join('|'))
```

The way I see it, this `set` method is going to be abstracted away by ComposeDB so the risk here is more a bad client implementation than a user/developer mistake and it's pretty limited, but if you feel strongly about it I can make the change?

Reply:
Yeah, I agree that for users going through ComposeDB this will be abstracted away for them, but I do think it's worth trying to design these APIs so they'll be easy to use for others building their own indexing solutions or trying to use the stream clients directly.

For a custom indexer, I guess it's probably true that they would likely have access to an instance of the corresponding Model and will be using the Model definition to figure out which fields to extract into the "unique" metadata anyway.  But if someone were building an app with the ModelInstanceDocument client directly without going through some other ComposeDB-like middleware, I could see the app just *knowing* which fields are the unique fields and providing them manually.

I'm conflicted, I do think most users are going to be going through some middleware like ComposeDB, OrbisDB, etc. Using the Stream clients directly is going to be rare.  That said, I do still feel nervous about client-facing APIs that are ordering-dependent based on some external context not necessarily available at the time that the API is being called...

I think I still lean slightly towards using a dictionary for this, but I do see your point and think the choice isn't as clear-cut as I first thought.  Maybe I'll tap in @ukstv and @dbcfd to see if either of them want to break the tie on this one.

Reply:
@dbcfd there's no way for our library to automatically sort or enforce sort ordering here, since our tooling doesn't have the context to know the app-specific meaning of the values being passed in.  That's why I suggested using a dictionary instead of an array where sort order matters, because then the developer is specifying the meaning of each argument explicitly, instead of implicitly via sort order

Reply:
@stbrody to be clear, I agree that at a high level we should use an object with explicit key/value pairs for the unique fields, that's what ComposeDB will expose in the runtime GraphQL schema, but there is a need to convert to an array of strings at some point.
So to me the question is more if we need to have structured object -> intermediary dictionary -> array of strings or if we can have structured object -> array of strings directly, which simplifies the implementation?

Reply:
lets put this into a ticket for future improvement i think.

Reply:
If we were to do this later it would be a breaking change.  I think if we want this to be the API we would need to change it now.